File: TechTest/app/src/main/java/com/david_kapas/android/techtest/presentation/posts/router/ListActivity.java
package com.david_kapas.android.techtest.presentation.posts.router;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;

import com.david_kapas.android.techtest.R;
import com.david_kapas.android.techtest.databinding.ActivityListBinding;
import com.david_kapas.android.techtest.di.modules.ListActivityModule;
import com.david_kapas.android.techtest.presentation.common.util.NetworkUtil;
import com.david_kapas.android.techtest.presentation.details.router.PostDetailsActivity;
import com.david_kapas.android.techtest.presentation.posts.model.PostListModel;
import com.david_kapas.android.techtest.presentation.posts.viewmodel.PostsViewModel;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

/**
 * Main screen for showing posts.
 * Created by David Kapas on 2018.03.17.
 */

public class ListActivity extends AppCompatActivity implements PostListRouter {

    @Inject
    PostsViewModel postsViewModel;
    @Inject
    ListActivityModule.ValamiFactory viewModelFactory;

    private boolean injected;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        ActivityListBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_list);
        PostListModel model = ViewModelProviders.of(this, viewModelFactory).get(PostListModel.class);
        binding.setViewModel(postsViewModel);
        if (NetworkUtil.checkNetworkConnection()) {
            postsViewModel.getAllPosts();
        } else {
            showGeneralErrorDialog();
        }
        //TODO: PullToRefresh not implemented due to short time frame

    }

    /*   private void inject() {
        if (!injected) {
            ListActivityComponent.Injector.buildComponent(this).inject(this);
            injected = true;
        }
    }*/

    @Override
    public void showGeneralErrorDialog() {
        //TODO: extended error handling not implemented due to short time frame
        showNetworkErrorDialog();
    }

    @Override
    public void openPost(int postId) {
        Intent intent = new Intent(this, PostDetailsActivity.class);
        intent.putExtra(PostDetailsActivity.POST_ID_PARAM, postId);
        startActivity(intent);
    }

    private void showNetworkErrorDialog() {
        AlertDialog alertDialog = new AlertDialog.Builder(ListActivity.this).create();
        alertDialog.setTitle(R.string.network_error_title);
        alertDialog.setMessage(getString(R.string.network_error_message));
        alertDialog.setButton(AlertDialog.BUTTON_NEUTRAL, "OK",
                (dialog, which) -> dialog.dismiss());
        alertDialog.show();
    }

}

File: TechTest/app/src/main/java/com/david_kapas/android/techtest/presentation/posts/router/ListActivity.kt
package com.david_kapas.android.techtest.presentation.posts.router

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import com.david_kapas.android.techtest.R
import com.david_kapas.android.techtest.databinding.ActivityListBinding
import com.david_kapas.android.techtest.di.modules.ListActivityModule
import com.david_kapas.android.techtest.presentation.common.util.NetworkUtil
import com.david_kapas.android.techtest.presentation.details.router.PostDetailsActivity
import com.david_kapas.android.techtest.presentation.posts.viewmodel.PostsViewModel
import dagger.android.AndroidInjection
import javax.inject.Inject

/**
 * Main screen for showing posts.
 * Created by David Kapas on 3/27/2019.
 */
class ListActivity : AppCompatActivity() {

    lateinit var postsViewModel: PostsViewModel
    @Inject
    lateinit var postsViewModelFactory: ListActivityModule.PostViewModelFactory
    lateinit var binding: ActivityListBinding;
    val alertDialog: AlertDialog by lazy { createAlertDialog() }

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        postsViewModel = ViewModelProviders.of(this, postsViewModelFactory).get(PostsViewModel::class.java);
        binding = DataBindingUtil.setContentView<ActivityListBinding>(this, R.layout.activity_list)
        binding.viewModel = postsViewModel
        postsViewModel.showErrorMessage.observe(this, Observer { showGeneralErrorDialog() })
        postsViewModel.openDetails.observe(this, Observer { openPost(it) })
        if (NetworkUtil.checkNetworkConnection()) {
            postsViewModel.getAllPosts()
        } else {
            showGeneralErrorDialog()
        }
        //TODO: PullToRefresh not implemented due to short time frame
    }

    override fun onDestroy() {
        alertDialog.dismiss()
        super.onDestroy()

    }

    fun showGeneralErrorDialog() {
        //TODO: extended error handling not implemented due to short time frame
        showNetworkErrorDialog()
    }

    fun openPost(it: Any?) {
        if (it != null) {
            var postId = it as Int
            val intent = Intent(this, PostDetailsActivity::class.java)
            intent.putExtra(PostDetailsActivity.POST_ID_PARAM, postId)
            startActivity(intent)
        }
    }

    private fun createAlertDialog(): AlertDialog {
        val alertDialog = AlertDialog.Builder(this@ListActivity).create()
        alertDialog.setTitle(R.string.network_error_title)
        alertDialog.setMessage(getString(R.string.network_error_message))
        alertDialog.setButton(AlertDialog.BUTTON_NEUTRAL, "OK"
        ) { dialog, which -> dialog.dismiss() }
        return alertDialog;
    }

    private fun showNetworkErrorDialog() {
        alertDialog.show()
    }
}